presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java

```diff
@@ -0,0 +1,235 @@
+        return node.accept(costEstimator, null);
+    }
+
+    private static class CostEstimator
+            extends PlanVisitor<PlanNodeCostEstimate, Void>
+    {
+        private final int numberOfNodes;
+        private final StatsProvider stats;
+        private final Lookup lookup;
+
+        CostEstimator(int numberOfNodes, StatsProvider stats, Lookup lookup)
+        {
+            this.numberOfNodes = numberOfNodes;
+            this.stats = requireNonNull(stats, "stats is null");
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
+        @Override
+        protected PlanNodeCostEstimate visitPlan(PlanNode node, Void context)
+        {
+            return UNKNOWN_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitGroupReference(GroupReference node, Void context)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitOutput(OutputNode node, Void context)
+        {
+            return ZERO_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitFilter(FilterNode node, Void context)
+        {
+            return cpuCost(getStats(node.getSource()).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitProject(ProjectNode node, Void context)
+        {
+            return cpuCost(getStats(node).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitAggregation(AggregationNode node, Void context)
+        {
+            PlanNodeStatsEstimate aggregationStats = getStats(node);
+            PlanNodeStatsEstimate sourceStats = getStats(node.getSource());
+            return new PlanNodeCostEstimate(sourceStats.getOutputSizeInBytes(), aggregationStats.getOutputSizeInBytes(), 0);
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitJoin(JoinNode node, Void context)
+        {
+            return calculateJoinCost(
+                    node,
+                    node.getLeft(),
+                    node.getRight(),
+                    node.getDistributionType().orElse(JoinNode.DistributionType.PARTITIONED).equals(JoinNode.DistributionType.REPLICATED));
+        }
+
+        private PlanNodeCostEstimate calculateJoinCost(PlanNode join, PlanNode probe, PlanNode build, boolean replicated)
+        {
+            int numberOfNodesMultiplier = replicated ? numberOfNodes : 1;
+
+            PlanNodeStatsEstimate probeStats = getStats(probe);
+            PlanNodeStatsEstimate buildStats = getStats(build);
+            PlanNodeStatsEstimate outputStats = getStats(join);
+
+            double cpuCost = probeStats.getOutputSizeInBytes() +
+                    buildStats.getOutputSizeInBytes() * numberOfNodesMultiplier +
+                    outputStats.getOutputSizeInBytes();
+
+            if (replicated) {
```

Comment:
no need for `if` as `1 - 1` gives `0` to add to `cpuCost`

Reply:
True, but it adds readability